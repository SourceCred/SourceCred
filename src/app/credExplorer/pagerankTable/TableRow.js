// @flow

import React, {type Node as ReactNode} from "react";

type TableRowProps = {|
  // How many layers of nested scope we're in (changes the color)
  +depth: number,
  // How many steps to indent the row (shifts button right)
  +indent: number,
  // The node that goes in the Description column
  +description: ReactNode,
  // What proportion should be formatted in the connection column
  +connectionProportion: ?number,
  // The score to format and display
  +score: number,
  // Children to show when the row is expanded
  +children: ReactNode,
|};
type TableRowState = {|
  expanded: boolean,
|};

export function scoreDisplay(score: number) {
  return score.toFixed(2);
}

export class TableRow extends React.PureComponent<
  TableRowProps,
  TableRowState
> {
  constructor() {
    super();
    this.state = {expanded: false};
  }
  render() {
    const {
      depth,
      indent,
      description,
      connectionProportion,
      score,
      children,
    } = this.props;
    const {expanded} = this.state;
    const percent =
      connectionProportion == null
        ? ""
        : (connectionProportion * 100).toFixed(2) + "%";
    return (
      <React.Fragment>
        <tr
          key="self"
          style={{backgroundColor: `rgba(0,143.4375,0,${1 - 0.9 ** depth})`}}
        >
          <td style={{display: "flex", alignItems: "flex-start"}}>
            <button
              style={{
                marginRight: 5,
                marginLeft: 15 * indent + 5,
              }}
              onClick={() => {
                this.setState(({expanded}) => ({
                  expanded: !expanded,
                }));
              }}
            >
              {expanded ? "\u2212" : "+"}
            </button>
            {description}
          </td>
          <td style={{textAlign: "right"}}>{percent}</td>
          <td style={{textAlign: "right"}}>{scoreDisplay(score)}</td>
        </tr>
        {expanded && children}
      </React.Fragment>
    );
  }
}
