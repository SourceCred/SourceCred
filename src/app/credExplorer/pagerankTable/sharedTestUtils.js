// @flow

import {dynamicAdapterSet} from "../../adapters/demoAdapters";
import {pagerank} from "../../../core/attribution/pagerank";
import {defaultWeightsForAdapterSet} from "../weights/weights";

export const COLUMNS = () => ["Description", "", "Cred"];

export async function example() {
  const adapters = await dynamicAdapterSet();
  const weightedTypes = defaultWeightsForAdapterSet(adapters.static());
  const graph = adapters.graph();
  const pnd = await pagerank(graph, (_unused_Edge) => ({
    toWeight: 1,
    froWeight: 1,
  }));

  return {adapters, pnd, weightedTypes};
}
