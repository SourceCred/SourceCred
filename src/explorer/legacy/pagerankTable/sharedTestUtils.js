// @flow

import React from "react";

import {type NodeAddressT} from "../../../core/graph";
import {pagerank} from "../../../analysis/pagerank";
import {graph as demoGraph} from "../../../plugins/demo/graph";
import {declaration as demoDeclaration} from "../../../plugins/demo/declaration";
import type {SharedProps} from "./shared";

export const COLUMNS = () => ["Description", "", "Cred"];

export async function example() {
  const graph = demoGraph();
  const declarations = [demoDeclaration];
  const pnd = await pagerank(graph, (_unused_Edge) => ({
    forwards: 1,
    backwards: 1,
  }));
  const maxEntriesPerList = 123;
  const manualWeights: Map<NodeAddressT, number> = new Map();
  const onManualWeightsChange: (NodeAddressT, number) => void = jest.fn();
  const weightConfig: any = <div data-test-weight-config={true} />;
  const weightFileManager: any = <div data-test-weight-file-manager={true} />;

  const sharedProps: SharedProps = {
    graph,
    declarations,
    pnd,
    maxEntriesPerList,
    manualWeights,
    onManualWeightsChange,
  };

  return {
    pnd,
    maxEntriesPerList,
    sharedProps,
    manualWeights,
    onManualWeightsChange,
    weightConfig,
    weightFileManager,
  };
}
