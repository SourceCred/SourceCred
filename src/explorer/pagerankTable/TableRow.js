// @flow

import React, {type Node as ReactNode} from "react";
import {StyleSheet, css} from "aphrodite/no-important";

type TableRowProps = {|
  // How many layers of nested scope we're in (changes the color)
  +depth: number,
  // How many steps to indent the row (shifts button right)
  +indent: number,
  // The node that goes in the Description column
  +description: ReactNode,

  // The content for the "multiuse column"
  // Could be a weight slider or a cred proportion depending on context.
  +multiuseColumn: ReactNode,
  // The cred amount to format and display
  +cred: number,
  // Children to show when the row is expanded
  +children: ReactNode,
  +showPadding: boolean,
|};
type TableRowState = {|
  expanded: boolean,
|};

export function credDisplay(cred: number) {
  return cred.toFixed(2);
}

export class TableRow extends React.PureComponent<
  TableRowProps,
  TableRowState
> {
  constructor() {
    super();
    this.state = {expanded: false};
  }
  render() {
    const {
      depth,
      indent,
      description,
      cred,
      children,
      showPadding,
      multiuseColumn,
    } = this.props;
    const {expanded} = this.state;
    const backgroundColor = `hsla(150,100%,28%,${1 - 0.9 ** depth})`;
    const makeGradient = (color) =>
      `linear-gradient(to top, ${color}, ${color})`;
    const normalBackground = makeGradient(backgroundColor);
    const highlightBackground = makeGradient("#D8E1E8");
    const backgroundImage = `${normalBackground}, ${highlightBackground}`;

    return (
      <React.Fragment>
        {showPadding && <PaddingRow backgroundColor={backgroundColor} />}
        <tr
          key="self"
          style={{backgroundImage}}
          className={css(styles.hoverHighlight)}
        >
          <td style={{display: "flex", alignItems: "flex-start"}}>
            <button
              style={{
                marginRight: 5,
                marginLeft: 15 * indent + 5,
              }}
              onClick={() => {
                this.setState(({expanded}) => ({
                  expanded: !expanded,
                }));
              }}
            >
              {expanded ? "\u2212" : "+"}
            </button>
            {description}
          </td>
          <td style={{textAlign: "right"}}>{multiuseColumn}</td>
          <td style={{textAlign: "right"}}>
            <span style={{marginRight: 5}}>{credDisplay(cred)}</span>
          </td>
        </tr>
        {expanded && children}
        {showPadding && <PaddingRow backgroundColor={backgroundColor} />}
      </React.Fragment>
    );
  }
}

export function PaddingRow(props: {|+backgroundColor: string|}) {
  return (
    <tr
      style={{
        height: 12,
        backgroundColor: props.backgroundColor,
      }}
    >
      <td />
      <td />
      <td />
    </tr>
  );
}

const styles = StyleSheet.create({
  /* To apply 'hoverHighlight', provide a backgroundImage containing two <image>
 * data types (eg linear gradients). The first backgroundImage will be
 * the default background. The second backgroundImage will be applied on top
 * of the first background when the user hovers or tabs over the element.
 */

  hoverHighlight: {
    backgroundSize: "100% 100%, 0 0",
    ":hover": {
      backgroundSize: "100% 100%, 100% 100%",
    },
    ":focus-within": {
      backgroundSize: "100% 100%, 100% 100%",
    },
  },
});
