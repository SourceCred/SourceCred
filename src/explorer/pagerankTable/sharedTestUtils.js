// @flow

import React from "react";

import {type NodeAddressT} from "../../core/graph";
import {pagerank} from "../../analysis/pagerank";
import {dynamicExplorerAdapterSet} from "../../plugins/demo/explorerAdapter";
import type {SharedProps} from "./shared";

export const COLUMNS = () => ["Description", "", "Cred"];

export async function example() {
  const adapters = await dynamicExplorerAdapterSet();
  const graph = adapters.graph();
  const pnd = await pagerank(graph, (_unused_Edge) => ({
    toWeight: 1,
    froWeight: 1,
  }));
  const maxEntriesPerList = 123;
  const manualWeights: Map<NodeAddressT, number> = new Map();
  const onManualWeightsChange: (NodeAddressT, number) => void = jest.fn();
  const weightConfig: any = <div data-test-weight-config={true} />;

  const sharedProps: SharedProps = {
    adapters,
    pnd,
    maxEntriesPerList,
    manualWeights,
    onManualWeightsChange,
  };

  return {
    adapters,
    pnd,
    maxEntriesPerList,
    sharedProps,
    manualWeights,
    onManualWeightsChange,
    weightConfig,
  };
}
