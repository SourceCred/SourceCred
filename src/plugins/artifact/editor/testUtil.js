// @flow

export function configureEnzyme() {
  const Enzyme = require("enzyme");
  const Adapter = require("enzyme-adapter-react-16");
  Enzyme.configure({adapter: new Adapter()});
}

export function configureAphrodite() {
  const {StyleSheetTestUtils} = require("aphrodite/no-important");
  beforeEach(() => {
    StyleSheetTestUtils.suppressStyleInjection();
  });

  afterEach(() => {
    StyleSheetTestUtils.clearBufferAndResumeStyleInjection();
  });
}
