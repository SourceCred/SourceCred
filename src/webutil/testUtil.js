// @flow

export function configureEnzyme() {
  const Enzyme = require("enzyme");
  const Adapter = require("enzyme-adapter-react-16");
  Enzyme.configure({adapter: new Adapter()});
  beforeEach(() => {
    // $FlowExpectedError
    console.error = jest.fn();
    // $FlowExpectedError
    console.warn = jest.fn();
  });
  afterEach(() => {
    expect(console.warn).not.toHaveBeenCalled();
    expect(console.error).not.toHaveBeenCalled();
  });
}

export function configureAphrodite() {
  const {StyleSheetTestUtils} = require("aphrodite/no-important");
  beforeEach(() => {
    StyleSheetTestUtils.suppressStyleInjection();
  });

  afterEach(() => {
    StyleSheetTestUtils.clearBufferAndResumeStyleInjection();
  });
}

export const relativeEntries = [
  {
    hash: "",
    pathname: "/foo/bar/",
    search: "",
    state: undefined,
  },
  {hash: "", pathname: "/1/", search: "", state: undefined},
  {hash: "", pathname: "/2/", search: "", state: undefined},
  {hash: "", pathname: "/3/", search: "", state: undefined},
  {hash: "", pathname: "/4/", search: "", state: undefined},
  {hash: "", pathname: "/5/", search: "", state: undefined},
];

export const memoryEntries = [
  {
    pathname: "/my/gateway/foo/bar/",
    search: "",
    hash: "",
    state: undefined,
  },
  {
    pathname: "/my/gateway/1/",
    search: "",
    hash: "",
    state: undefined,
  },
  {
    pathname: "/my/gateway/2/",
    search: "",
    hash: "",
    state: undefined,
  },
  {
    pathname: "/my/gateway/3/",
    search: "",
    hash: "",
    state: undefined,
  },
  {
    pathname: "/my/gateway/4/",
    search: "",
    hash: "",
    state: undefined,
  },
  {
    pathname: "/my/gateway/5/",
    search: "",
    hash: "",
    state: undefined,
  },
];
