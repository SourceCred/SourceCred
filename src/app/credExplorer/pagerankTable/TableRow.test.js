// @flow

import React from "react";
import {shallow} from "enzyme";
import {TableRow} from "./TableRow";

import {COLUMNS} from "./sharedTestUtils";
require("../../testUtil").configureEnzyme();

describe("app/credExplorer/pagerankTable/TableRow", () => {
  function example() {
    return shallow(
      <TableRow
        depth={1}
        indent={1}
        description={<span data-test-description={true} />}
        connectionProportion={0.5}
        score={133.7}
        children={<div data-test-children={true} />}
      />
    );
  }
  it("depth parameter changes the color, but not the indentation", () => {
    for (const depth of [0, 1, 2]) {
      const el = shallow(
        <TableRow
          depth={depth}
          indent={1}
          description={<span data-test-description={true} />}
          connectionProportion={0.5}
          score={133.7}
          children={<div data-test-children={true} />}
        />
      );
      const trStyle = el.find("tr").props().style;
      const buttonStyle = el.find("button").props().style;
      expect({depth, trStyle, buttonStyle}).toMatchSnapshot();
    }
  });
  it("indent parameter changes the button indentation", () => {
    for (const indent of [0, 1, 2]) {
      const el = shallow(
        <TableRow
          depth={3}
          indent={indent}
          description={<span data-test-description={true} />}
          connectionProportion={0.5}
          score={133.7}
          children={<div data-test-children={true} />}
        />
      );
      const trStyle = el.find("tr").props().style;
      const buttonStyle = el.find("button").props().style;
      expect({indent, trStyle, buttonStyle}).toMatchSnapshot();
    }
  });
  it("expand button toggles symbol based on expansion state", () => {
    const el = example();
    el.setState({expanded: false});
    expect(el.find("button").text()).toEqual("+");
    el.setState({expanded: true});
    expect(el.find("button").text()).toEqual("\u2212");
  });
  it("clicking the expand button toggles expansion state", () => {
    const el = example();
    el.setState({expanded: false});
    el.find("button").simulate("click");
    expect(el.state().expanded).toBe(true);
    el.find("button").simulate("click");
    expect(el.state().expanded).toBe(false);
  });
  it("defaults to not expanded", () => {
    const el = example();
    expect(el.state().expanded).toBe(false);
  });
  it("displays children only when expanded", () => {
    const el = example();
    el.setState({expanded: false});
    expect(el.find({"data-test-children": true})).toHaveLength(0);
    el.setState({expanded: true});
    expect(el.find({"data-test-children": true})).toHaveLength(1);
  });
  it("has the correct number of columns", () => {
    const el = example();
    expect(el.find("td")).toHaveLength(COLUMNS().length);
  });
  it("displays formatted connectionPercentage in the correct column", () => {
    const index = COLUMNS().indexOf("Connection");
    expect(index).not.toEqual(-1);
    const td = example()
      .find("td")
      .at(index);
    expect(td.text()).toEqual("50.00%");
  });
  it("displays empty column when connectionProportion not set", () => {
    const index = COLUMNS().indexOf("Connection");
    expect(index).not.toEqual(-1);
    const el = example();
    el.setProps({connectionProportion: null});
    const td = el.find("td").at(index);
    expect(td.text()).toEqual("");
  });
  it("displays formatted score in the correct column", () => {
    const index = COLUMNS().indexOf("Score");
    expect(index).not.toEqual(-1);
    const td = example()
      .find("td")
      .at(index);
    expect(td.text()).toEqual("133.70");
  });
  it("displays the description in the correct column", () => {
    const index = COLUMNS().indexOf("Description");
    expect(index).not.toEqual(-1);
    const td = example()
      .find("td")
      .at(index);
    expect(td.find({"data-test-description": true})).toHaveLength(1);
  });
});
