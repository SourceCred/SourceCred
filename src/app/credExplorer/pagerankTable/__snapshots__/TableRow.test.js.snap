// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/credExplorer/pagerankTable/TableRow depth parameter changes the color, but not the indentation 1`] = `
Object {
  "buttonStyle": Object {
    "marginLeft": 20,
    "marginRight": 5,
  },
  "depth": 0,
  "trStyle": Object {
    "backgroundColor": "rgba(0,143.4375,0,0)",
  },
}
`;

exports[`app/credExplorer/pagerankTable/TableRow depth parameter changes the color, but not the indentation 2`] = `
Object {
  "buttonStyle": Object {
    "marginLeft": 20,
    "marginRight": 5,
  },
  "depth": 1,
  "trStyle": Object {
    "backgroundColor": "rgba(0,143.4375,0,0.09999999999999998)",
  },
}
`;

exports[`app/credExplorer/pagerankTable/TableRow depth parameter changes the color, but not the indentation 3`] = `
Object {
  "buttonStyle": Object {
    "marginLeft": 20,
    "marginRight": 5,
  },
  "depth": 2,
  "trStyle": Object {
    "backgroundColor": "rgba(0,143.4375,0,0.18999999999999995)",
  },
}
`;

exports[`app/credExplorer/pagerankTable/TableRow indent parameter changes the button indentation 1`] = `
Object {
  "buttonStyle": Object {
    "marginLeft": 5,
    "marginRight": 5,
  },
  "indent": 0,
  "trStyle": Object {
    "backgroundColor": "rgba(0,143.4375,0,0.2709999999999999)",
  },
}
`;

exports[`app/credExplorer/pagerankTable/TableRow indent parameter changes the button indentation 2`] = `
Object {
  "buttonStyle": Object {
    "marginLeft": 20,
    "marginRight": 5,
  },
  "indent": 1,
  "trStyle": Object {
    "backgroundColor": "rgba(0,143.4375,0,0.2709999999999999)",
  },
}
`;

exports[`app/credExplorer/pagerankTable/TableRow indent parameter changes the button indentation 3`] = `
Object {
  "buttonStyle": Object {
    "marginLeft": 35,
    "marginRight": 5,
  },
  "indent": 2,
  "trStyle": Object {
    "backgroundColor": "rgba(0,143.4375,0,0.2709999999999999)",
  },
}
`;
